### ISSUE TITLE
BUG: MultiIndex.set_names loses name

### ISSUE THREAD

**Issue by TomAugspurger** [pandas-dev/pandas#11979]
Seems surprising, I thought `.name` would come along. 

``` python
In [31]: idx = pd.MultiIndex.from_product([['A'], ['a', 'b', 'c']])

In [32]: idx.name = 'foo'

In [33]: idx2 = idx.set_names(['l1', 'l2'])

In [34]: idx2.name
```

Bug, or user error since MultiIndex.name isn't all that useful?

**Comment by jreback** [comment]
I think the fix is similar to what I am doing in ``RangeIndex``, see here: https://github.com/pydata/pandas/pull/11892/files#diff-dc96e62847cdfe77b570cf9c6b0e8086R495

``name`` and ``names`` are accepted and aliased to each other for both ``Index`` and ``MultiIndex``. mainly for compat I think (e.g. for an ``Index`` people tend to use ``name`` and for a ``MultiIndex`` names). You can't have both, but either should set correctly. (of course a ``MultiIndex`` index requires a list equal in len to the levels; ``.set_names`` enforces this). I think the setter needs to be updated though.

**Comment by jreback** [comment]
does #11980 close this?

**Comment by TomAugspurger** [comment]
No, I think I had an old test would have failed, but I removed the ability to set both `self.index.name` and `self.index.names` on a MultiIndex

**Comment by jreback** [comment]
thanks @TomAugspurger

**Comment by SeeminSyed** [comment]
Does #30335 close this ticket and if not, would I be possible for me to work on it?

**Comment by SeeminSyed** [comment]
@TomAugspurger @jreback @jbrockmendel ^^^

Sorry to disturb but I'm not sure notifications are sent if people are not mentioned by name.

**Comment by TomAugspurger** [comment]
No, this isn't fixed by #30335.

**Comment by jbrockmendel** [comment]
Allowing MultiIndex.name to be set seems more likely to cause confusion than anything else.  I say we deprecate.
